src/sinks/splunk_hec/common/acknowledgements.rs

```diff
@@ -0,0 +1,314 @@
+use super::service::HttpRequestBuilder;
+use crate::http::HttpClient;
+use hyper::Body;
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, num::NonZeroU8, sync::Arc, time::Duration};
+use tokio::sync::{mpsc::UnboundedReceiver, oneshot::Sender};
+use vector_core::event::EventStatus;
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct HecClientAcknowledgementsConfig {
+    pub query_interval: NonZeroU8,
+    pub retry_limit: NonZeroU8,
+}
+
+impl Default for HecClientAcknowledgementsConfig {
+    fn default() -> Self {
+        Self {
+            query_interval: NonZeroU8::new(10).unwrap(),
+            retry_limit: NonZeroU8::new(30).unwrap(),
+        }
+    }
+}
+
+pub fn default_hec_client_acknowledgements_config() -> Option<HecClientAcknowledgementsConfig> {
+    Some(Default::default())
+}
+
+#[derive(Deserialize, Serialize, Eq, PartialEq, Debug)]
+pub struct HecAckStatusRequest {
+    pub acks: Vec<u64>,
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+pub struct HecAckStatusResponse {
+    pub acks: HashMap<u64, bool>,
+}
+
+enum HecAckApiError {
+    ClientBuildRequest,
+    ClientParseResponse,
+    ClientSendQuery,
+    ServerSendQuery,
+}
+
+struct HecAckClient {
+    acks: HashMap<u64, (u8, Sender<EventStatus>)>,
+    retry_limit: u8,
+    client: HttpClient,
+    http_request_builder: Arc<HttpRequestBuilder>,
+}
+
+impl HecAckClient {
+    fn new(
+        retry_limit: u8,
+        client: HttpClient,
+        http_request_builder: Arc<HttpRequestBuilder>,
+    ) -> Self {
+        Self {
+            acks: HashMap::new(),
+            retry_limit,
+            client,
+            http_request_builder,
+        }
+    }
+
+    /// Adds an ack id to be queried
+    fn add(&mut self, ack_id: u64, ack_event_status_sender: Sender<EventStatus>) {
+        self.acks
+            .insert(ack_id, (self.retry_limit, ack_event_status_sender));
+    }
+
+    /// Queries Splunk HEC with stored ack ids and finalizes events that are successfully acked
+    async fn run(&mut self) {
+        let ack_query_body = self.get_ack_query_body();
+        if !ack_query_body.acks.is_empty() {
+            let ack_query_response = self.send_ack_query_request(&ack_query_body).await;
+
+            match ack_query_response {
+                Ok(ack_query_response) => {
+                    debug!("Received ack statuses {:?}", ack_query_response);
```

Comment:
Usually we like to separate out fields, even for debug messages. This would be something like:

```rust
debug!(message = "Received ack statuses", ?ack_query_response);
```